controller/api/destination/watcher/profile_watcher.go

```diff
@@ -0,0 +1,234 @@
+		},
+	)
+
+	return watcher
+}
+
+//////////////////////
+/// ProfileWatcher ///
+//////////////////////
+
+// Subscribe to an authority.
+// The provided listener will be updated each time the service profile for the
+// given authority is changed.
+func (pw *ProfileWatcher) Subscribe(authority string, contextToken string, listener ProfileUpdateListener) error {
+	name, err := profileID(authority, contextToken)
+	if err != nil {
+		return err
+	}
+
+	pw.log.Infof("Establishing watch on profile %s", name)
+
+	pw.profilesMu.Lock()
+	publisher, ok := pw.profiles[name]
+	if !ok {
+		profile, err := pw.profileLister.ServiceProfiles(name.Namespace).Get(name.Name)
+		if err != nil {
+			profile = nil
+		}
+
+		publisher = pw.newProfilePublisher(name, profile)
+		pw.profiles[name] = publisher
+	}
+	pw.profilesMu.Unlock()
+
+	publisher.subscribe(listener)
+	return nil
+}
+
+// Unsubscribe removes a listener from the subscribers list for this authority.
+func (pw *ProfileWatcher) Unsubscribe(authority string, contextToken string, listener ProfileUpdateListener) error {
+	name, err := profileID(authority, contextToken)
+	if err != nil {
+		return err
+	}
+	pw.log.Infof("Stopping watch on profile %s", name)
+
+	pw.profilesMu.RLock()
+	publisher, ok := pw.profiles[name]
+	pw.profilesMu.RUnlock()
+	if !ok {
+		return fmt.Errorf("cannot unsubscribe from unknown service [%s] ", name)
+	}
+	publisher.unsubscribe(listener)
+	return nil
+}
+
+func (pw *ProfileWatcher) addProfile(obj interface{}) {
+	profile := obj.(*sp.ServiceProfile)
+	id := ProfileID{
+		Namespace: profile.Namespace,
+		Name:      profile.Name,
+	}
+
+	pw.profilesMu.Lock()
+	publisher, ok := pw.profiles[id]
+	if !ok {
+		publisher = pw.newProfilePublisher(id, profile)
+		pw.profiles[id] = publisher
+
+	}
+	pw.profilesMu.Unlock()
+
+	publisher.update(profile)
+}
+
+func (pw *ProfileWatcher) updateProfile(old interface{}, new interface{}) {
+	pw.addProfile(new)
+}
+
+func (pw *ProfileWatcher) deleteProfile(obj interface{}) {
+	profile := obj.(*sp.ServiceProfile)
+	id := ProfileID{
+		Namespace: profile.Namespace,
+		Name:      profile.Name,
+	}
+
+	pw.profilesMu.RLock()
+	publisher, ok := pw.profiles[id]
+	pw.profilesMu.RUnlock()
+	if ok {
+		publisher.update(nil)
+	}
+}
+
+func (pw *ProfileWatcher) newProfilePublisher(id ProfileID, profile *sp.ServiceProfile) *profilePublisher {
+	return &profilePublisher{
+		profile:   profile,
+		listeners: make([]ProfileUpdateListener, 0),
+		log: pw.log.WithFields(logging.Fields{
+			"component": "profile-publisher",
+			"ns":        id.Namespace,
+			"profile":   id.Name,
+		}),
+	}
+}
+
+////////////////////////
+/// profilePublisher ///
+////////////////////////
+
+func (pp *profilePublisher) subscribe(listener ProfileUpdateListener) {
+	pp.mutex.Lock()
+	defer pp.mutex.Unlock()
+
+	pp.listeners = append(pp.listeners, listener)
+	listener.Update(pp.profile)
+}
+
+// unsubscribe returns true iff the listener was found and removed.
```

Comment:
iff -> if

Reply:
Old mathie habits die hard (https://www.merriam-webster.com/dictionary/iff).  I'll update this to human English.